GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/IPC/VFSForGitNotification.m

```diff
@@ -0,0 +1,142 @@
+    }
+    
+    Identifier idValue = [identifier integerValue];
+    switch (idValue)
+    {
+        case AutomountStart:
+        {
+            *notification = [[VFSForGitNotification alloc]
+                             initAsAutomountStartWithMessage:jsonMessage
+                             error:error];
+            break;
+        }
+        
+        case MountSuccess:
+        {
+            *notification = [[VFSForGitNotification alloc]
+                             initAsMountSuccessWithMessage:jsonMessage
+                             error:error];
+            break;
+        }
+            
+        case MountFailure:
+        {
+            *notification = [[VFSForGitNotification alloc]
+                             initAsMountFailureWithMessage:jsonMessage
+                             error:error];
+            break;
+        }
+            
+        default:
+        {
+            *notification = nil;
+            *error = @"Unrecognised message id";
+            break;
+        }
+    }
+    
+    return *notification != nil;
+}
+
+- (instancetype) initAsAutomountStartWithMessage:(NSDictionary *_Nonnull) messageDict
+                                           error:(NSString *_Nonnull *_Nonnull) error
+{
+    if (self = [super init])
+    {
+        id repoCount = messageDict[EnlistmentCountKey];
+        if (repoCount && [repoCount isKindOfClass:[NSNumber class]])
+        {
+            _title = [(NSString *) AutomountTitle copy];
+            _message = [[NSString stringWithFormat:
+                         (NSString *) AutomountStartMessageFormat,
+                         [repoCount unsignedIntegerValue]] copy];
+            return self;
+        }
+        
+        *error = @"Missing repos count in AutomountStart message";
+        self = nil;
+    }
+    
+    return self;
+}
+
+- (instancetype) initAsMountSuccessWithMessage:(NSDictionary *_Nonnull) messageDict
+                                         error:(NSString *_Nonnull *_Nonnull) error
+{
+    return self =
+    [self initAsMountWithMessage:messageDict
+                           title:(NSString *) AutomountTitle
+                   messageFormat:(NSString *) AutomountSuccessMessageFormat
+                           error:error];
+}
+
+- (instancetype) initAsMountFailureWithMessage:(NSDictionary *_Nonnull) messageDict
+                                         error:(NSString *_Nonnull *_Nonnull) error
+{
+    return self =
+    [self initAsMountWithMessage:messageDict
+                           title:(NSString *) AutomountTitle
+                   messageFormat:(NSString *) AutomountFailureMessageFormat
+                           error:error];
+}
+
+- (instancetype) initAsMountWithMessage:(NSDictionary *_Nonnull) messageDict
+                                  title:(NSString *_Nonnull) title
+                          messageFormat:(NSString *_Nonnull) messageFormat
+                                  error:(NSString *_Nonnull *_Nonnull) error
```

Comment:
Doesn't "*_Nonnull *_Nonnull" imply error!=nil && (*error)!=nil? That doesn't look right for an output parameter

Reply:
It does look confusing. Couple references - 
1. [Apple documentation](https://developer.apple.com/swift/blog/?id=25). See the line (although it does not exactly match our case) ```More complex pointer types like id * must be explicitly annotated. For example, to specify a non-nullable pointer to a nullable object reference, use _Nullable id * _Nonnull.```
2. Did search in Apple frameworks, could find a few sample code
```
MPSKeyedArchiver.h

+ (nullable id)unarchiveTopLevelObjectWithData:(NSData *__nonnull)data
                                         error:(NSError * __nullable * __nullable)error NS_UNAVAILABLE;
```

```
MTKModel.h

- (nullable instancetype)initWithMesh:(nonnull MDLMesh *)mesh
                               device:(nonnull id<MTLDevice>)device
                                error:(NSError * __nullable * __nullable)error;
```

Reply:
Let me know if there is a more readable alternative...

Reply:
Those are all "nullable" decorations, you are using the opposite "nonnull" decoration. Both error and *error can be null/nil, so you would use __nullable decoration for that.

Reply:
makes sense. Updated to use the default nullability for `NSError`s.